services/src/machine_learning/mod.rs

```diff
@@ -0,0 +1,107 @@
+mod data_preparation;
+mod ml_tasks;
+pub mod xg_error;
+pub mod xgb_config_builder;
+mod xgboost_training;
+
+pub(crate) use ml_tasks::{schedule_ml_model_training_task, MLTrainRequest};
+
+#[cfg(test)] //TODO: remove test config, once its used outside of tests
+pub(crate) use ml_tasks::MachineLearningModelFromWorkflowResult;
+
+use serde::Deserialize;
+use serde::Serialize;
+use typetag::serde;
+pub(crate) use xgboost_training::xgb_train_model;
+pub(crate) use xgboost_training::XgboostTrainingParams;
+
+/// Represents a single feature of a machine learning training data set
+#[derive(Debug, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct MachineLearningFeature {
+    pub feature_name: Option<String>, // e.g. temperature, elevation etc.
+    pub feature_data: Vec<f32>,
+}
+
+impl MachineLearningFeature {
+    pub fn new(feature_name: Option<String>, feature_data: Vec<f32>) -> MachineLearningFeature {
+        MachineLearningFeature {
+            feature_name,
+            feature_data,
+        }
+    }
+}
+
+/// This enum represents the different aggregators that can be used to initialize the
+/// different algorithms for collecting the data used in ml training.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+#[serde(rename_all = "camelCase")]
+pub enum MachineLearningAggregator {
+    Simple,
+    ReservoirSampling,
+}
+
+/// The purpose of this trait is to facilitate a generic aggregation process of raster data.
+pub trait Aggregatable {
+    type Data;
+
+    /// This method should realize the aggregation algorithm of the implementing struct.
+    fn aggregate(&mut self, incoming_data: &mut dyn Iterator<Item = Option<Self::Data>>);
+
+    /// Once the aggregation process is finished, return the data for further usage.
+    fn finish(&mut self) -> Vec<Self::Data>;
+
+    /// Used to generate a new instance of an aggregatable within a generic context
+    fn new(data: Vec<f32>, memory_limit: Option<usize>) -> Self;
+}
+
+/// A simple aggregator that just collects all incoming data in a vector.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct SimpleAggregator {
```

Comment:
if you use `data: <Vec<T>>` you can implement the `Aggregatable` trait in a generic way.

Reply:
The output type of the aggregatable should be f32 in any case (at least for now, because xgboost wants f32 inputs).
So maybe this change is ok, too? Now only the aggregate function takes a generic type that can be converted to a f32 vector.

ff40127358edaba628f74a6b6d0d0f0f2c8630d3

---
commit msg:
refactor: make the aggregate function of the aggregatable trait generic over the input iterator associated type